fili-core/src/main/java/com/yahoo/bard/webservice/data/metric/protocol/TimeAverageMetricTransformer.java

```diff
@@ -0,0 +1,215 @@
+import com.yahoo.bard.webservice.data.metric.MetricDictionary;
+import com.yahoo.bard.webservice.data.time.ZonelessTimeGrain;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * An interface for transforming metrics into other metrics.
+ */
+public class TimeAverageMetricTransformer implements MetricTransformer {
+
+    private static final MetricDictionary EMPTY_METRIC_DICTIONARY = new MetricDictionary();
+
+    public static final String CONTRACT_NAME = "reagg";
+    private static final String CORE_PARAMETER = "reagg";
+
+    public final static TimeAverageMetricTransformer INSTANCE = new TimeAverageMetricTransformer();
+
+    private static final String NAME_FORMAT = "%s%s";
+    private static final String LONG_NAME_FORMAT = "%s (%s)";
+    private static final String DESCRIPTION_FORMAT = "The %s of %s.";
+
+    private Map<String, TimeAverageMetricMakerConfig> makerConfigMap;
+
+    private Map<String, AggregationAverageMaker> metricMakerMap;
+    private final Supplier<ProtocolSupport> protocolSupportSupplier;
+
+    /**
+     * Constructor.
+     *
+     * Left protected to allow subclasses to add custom time intervals for reaggregation.
+     *
+     * @param protocolSupportSupplier  A source for the default protocol support used by the makers.
+     * @param  makerConfigMap  A collection of maker configurations to use when making time reaggregations.
+     */
+    protected TimeAverageMetricTransformer(
+            Supplier<ProtocolSupport> protocolSupportSupplier,
+            Map<String, TimeAverageMetricMakerConfig> makerConfigMap
+    ) {
+        this.protocolSupportSupplier = protocolSupportSupplier;
+        this.makerConfigMap = makerConfigMap;
+        metricMakerMap = new HashMap<>();
+    }
+
+    /**
+     * Constructor.
+     *
+     * Private to implement singleton pattern for normal usage.
+     */
+    private TimeAverageMetricTransformer() {
+        this(BuiltInMetricProtocols::getStandardProtocolSupport, TimeAverageMetricMakerConfig.timeMakerConfigs);
+    }
+
+    /**
+     * Build a MetricMaker for this configuration.
+     *
+     * @param makerConfig  The configuration for the metric maker.
+     *
+     * @return  A metric maker for this configuration.
+     */
+    private AggregationAverageMaker buildMaker(TimeAverageMetricMakerConfig makerConfig) {
+        return new AggregationAverageMaker(
+                EMPTY_METRIC_DICTIONARY,
+                makerConfig.getGrain(),
+                protocolSupportSupplier.get()
+        );
+    }
+    @Override
+    public LogicalMetric apply(LogicalMetric logicalMetric, Protocol protocol, Map<String, String> parameterValues)
+            throws UnknownProtocolValueException {
+
+        String parameterValue = parameterValues.get(CORE_PARAMETER);
+
+        if (!makerConfigMap.containsKey(parameterValue)) {
```

Comment:
Two things:
1. We need to check if `parameterValue` is null. `Map` implementations that do not support null keys will throw a null pointer on `Map#containsKey` (https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#containsKey-java.lang.Object-)
2. I think if you are explicitly tying this Transformer to the "reagg" core parameter, the CORE_PARAMETER must be `public`. Currently it is `private`.